processing/src/main/java/io/druid/query/dimension/LookupDimensionSpec.java

```diff
@@ -0,0 +1,251 @@
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.dimension;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.metamx.common.StringUtils;
+import io.druid.query.extraction.DimExtractionFn;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.query.extraction.LookupExtractor;
+import io.druid.query.extraction.LookupReferencesManager;
+import io.druid.query.filter.DimFilterCacheHelper;
+import io.druid.segment.DimensionSelector;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+public class LookupDimensionSpec implements DimensionSpec
+{
+  private static final byte CACHE_TYPE_ID = 0x4;
+
+  @JsonProperty
+  private final String dimension;
+
+  @JsonProperty
+  private final String outputName;
+
+  @JsonProperty
+  private final LookupExtractor lookup;
+
+  @JsonProperty
+  private final boolean retainMissingValues;
+
+  @JsonProperty
+  private final String replaceMissingWith;
+
+  @JsonProperty
+  private final String name;
+
+
+  private final LookupReferencesManager lookupReferencesManager;
+
+
+  @JsonCreator
+  public LookupDimensionSpec(
+      @JsonProperty("dimension") String dimension,
+      @JsonProperty("outputName") String outputName,
+      @JsonProperty("lookup") LookupExtractor lookup,
+      @JsonProperty("retainMissingValues") final boolean retainMissingValues,
+      @JsonProperty("replaceMissingWith") final String replaceMissingWith,
+      @JsonProperty("name") String name,
+      @JacksonInject LookupReferencesManager lookupReferencesManager
+  )
+  {
+    this.retainMissingValues = retainMissingValues;
+    this.replaceMissingWith = Strings.emptyToNull(replaceMissingWith);
+    Preconditions.checkArgument(
+        !(retainMissingValues && !Strings.isNullOrEmpty(replaceMissingWith)),
+        "Cannot specify a [replaceMissingValue] and set [retainMissingValue] to true"
+    );
+    this.dimension = Preconditions.checkNotNull(dimension, "dimension can not be Null");
+    this.outputName = Preconditions.checkNotNull(outputName, "outputName can not be Null");
+    this.lookupReferencesManager = lookupReferencesManager;
+    this.name = name;
+    this.lookup = lookup;
+    Preconditions.checkArgument(
+        Strings.isNullOrEmpty(name) ^ (lookup == null),
+        "name and lookup are mutually exclusive please provide either a name or a lookup"
```

Comment:
it would probably be nice to return what was passed in as well.  IIRC, `Preconditions` can take `String.format()` style strings, so it should be fine to add "%s" and then pass along name and lookup.